plugins/builtin/target/gce-mig/plugin/gce.go

```diff
@@ -0,0 +1,157 @@
+package plugin
+
+import (
+	"context"
+	"fmt"
+	"github.com/hashicorp/nomad-autoscaler/sdk"
+	"github.com/hashicorp/nomad-autoscaler/sdk/helper/scaleutils"
+	"github.com/mitchellh/go-homedir"
+	"google.golang.org/api/compute/v1"
+	"google.golang.org/api/option"
+	"io/ioutil"
+	"os"
+	"time"
+)
+
+func (t *TargetPlugin) setupGCEClients(config map[string]string) error {
+
+	credentials, ok := config[configKeyCredentials]
+
+	if ok {
+		contents, err := pathOrContents(credentials)
+		if err != nil {
+			return fmt.Errorf("failed to read credentials: %v", err)
+		}
+
+		t.service, err = compute.NewService(context.Background(), option.WithCredentialsJSON([]byte(contents)))
+
+		if err != nil {
+			return fmt.Errorf("failed to create Google Compute Engine client: %v", err)
+		}
+	} else {
+		service, err := compute.NewService(context.Background())
+
+		if err != nil {
+			return fmt.Errorf("failed to create Google Compute Engine client: %v", err)
+		}
+
+		t.service = service
+	}
+
+	return nil
+}
+
+func (t *TargetPlugin) scaleOut(ctx context.Context, project string, zone string, mig string, num int64) error {
+
+	_, err := t.service.RegionInstanceGroupManagers.Resize(project, zone, mig, num).Context(ctx).Do()
+
+	if err != nil {
+		return fmt.Errorf("failed to scale out instance group: %v", err)
+	}
+
+	return nil
+}
+
+func (t *TargetPlugin) scaleIn(ctx context.Context, project string, region string, mig string, num int64, config map[string]string) error {
+	scaleReq, err := t.generateScaleReq(num, config)
+	if err != nil {
+		return fmt.Errorf("failed to generate scale in request: %v", err)
+	}
+
+	ids, err := t.scaleInUtils.RunPreScaleInTasks(ctx, scaleReq)
+	if err != nil {
+		return fmt.Errorf("failed to perform pre-scale Nomad scale in tasks: %v", err)
+	}
+
+	// Grab the instanceIDs once as it is used multiple times throughout the
+	// scale in event.
+	var instanceIDs []string
+
+	for _, node := range ids {
+		instanceIDs = append(instanceIDs, node.RemoteID)
+	}
+
+	// Create a logger for this action to pre-populate useful information we
+	// would like on all log lines.
+	log := t.logger.With("action", "scale_in", "instance_group", mig, "instances", ids)
+
+	// Terminate the detached instances.
```

Comment:
Could this maybe be reworded? My understanding is that the instances do not need to be detached before calling terminate and that is all handled as part of the `DeleteInstances` call?